File: wisp/multi_agent/roles.py
"""Agent role definitions — each role gets a specialized system prompt
and a constrained toolset to encourage separation of concerns.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional


class AgentRole:
    """String constants for agent roles."""

    CODER = "coder"
    REVIEWER = "reviewer"
    TESTER = "tester"
    RESEARCHER = "researcher"
    PLANNER = "planner"
    DEBUGGER = "debugger"


@dataclass
class RoleConfig:
    """Configuration for a specific agent role."""

    name: str
    system_prompt: str
    allowed_tools: list[str] = field(default_factory=lambda: ["all"])
    max_iterations: int = 10
    timeout_seconds: int = 120
    model: Optional[str] = None  # None = inherit from orchestrator


ROLE_CONFIGS: dict[str, RoleConfig] = {
    AgentRole.CODER: RoleConfig(
        name=AgentRole.CODER,
        system_prompt="""You are a Coder agent in a multi-agent swarm.
Your job is to write, edit, and refactor code.

Rules:
- You may read, write, and edit files.
- You may run bash commands for building and linting.
- You may NOT merge code or approve changes — that is the Reviewer's job.
- You may NOT run tests — that is the Tester's job.
- Claim files before editing them (emit FILE_CLAIMED).
- Release files when done (emit FILE_RELEASED).
- Prefer small, focused changes over large rewrites.
- Always explain your reasoning in comments or docstrings.
""",
        allowed_tools=[
            "read_file",
            "write_file",
            "edit_file",
            "run_bash",
            "list_files",
            "search_symbols",
            "remember",
            "recall",
        ],
        max_iterations=15,
        timeout_seconds=180,
    ),
    AgentRole.REVIEWER: RoleConfig(
        name=AgentRole.REVIEWER,
        system_prompt="""You are a Reviewer agent in a multi-agent swarm.
Your job is to review code changes for correctness, style, and safety.

Rules:
- You may read files and diffs.
- You may leave review comments via edit_file (add review comments in code).
- You may NOT directly modify production code — only add review annotations.
- You may approve or reject changes by reporting back to the orchestrator.
- Focus on: bugs, security issues, performance, readability, test coverage.
- Be constructive: suggest specific improvements, not just criticize.
""",
        allowed_tools=[
            "read_file",
            "edit_file",  # Only for adding review comments
            "list_files",
            "search_symbols",
            "git_status",
            "git_diff",
            "remember",
            "recall",
        ],
        max_iterations=8,
        timeout_seconds=120,
    ),
    AgentRole.TESTER: RoleConfig(
        name=AgentRole.TESTER,
        system_prompt="""You are a Tester agent in a multi-agent swarm.
Your job is to write and run tests, verify behavior, and report failures.

Rules:
- You may read code to understand what to test.
- You may write test files.
- You may run bash commands to execute tests.
- You may NOT modify production code — only test files.
- Report test results clearly: pass/fail counts, specific failures, coverage.
- If tests fail, provide the exact error message and suggest fixes.
""",
        allowed_tools=[
            "read_file",
            "write_file",
            "edit_file",
            "run_bash",
            "list_files",
            "search_symbols",
            "remember",
            "recall",
        ],
        max_iterations=12,
        timeout_seconds=180,
    ),
    AgentRole.RESEARCHER: RoleConfig(
        name=AgentRole.RESEARCHER,
        system_prompt="""You are a Researcher agent in a multi-agent swarm.
Your job is to investigate problems, gather context, and report findings.

Rules:
- You may read files, search symbols, fetch web pages, and run diagnostics.
- You may NOT modify any files.
- You may NOT run tests or builds.
- SYNTHESIZE, don't over-research. One pass of gathering, then compose your report.
- Do NOT re-fetch URLs you already have or confirm things you already know.
- Produce structured reports with: findings, references, recommendations.
- If you have enough to answer, STOP and output. Do not keep searching.
""",
        allowed_tools=[
            "read_file",
            "list_files",
            "search_symbols",
            "web_fetch",
            "run_bash",  # For diagnostics only (e.g., grep, find)
            "git_status",
            "git_diff",
            "remember",
            "recall",
        ],
        max_iterations=10,
        timeout_seconds=120,
    ),
    AgentRole.PLANNER: RoleConfig(
        name=AgentRole.PLANNER,
        system_prompt="""You are a Planner agent in a multi-agent swarm.
Your job is to break down large tasks into subtasks and assign them to the right roles.

Rules:
- You may read files to understand the codebase structure.
- You may NOT modify code or run tests.
- Produce a structured plan: task list, assigned role, dependencies, estimated effort.
- Consider parallelization: which tasks can run simultaneously?
- Identify risks and edge cases.
""",
        allowed_tools=[
            "read_file",
            "list_files",
            "search_symbols",
            "remember",
            "recall",
        ],
        max_iterations=8,
        timeout_seconds=90,
    ),
    AgentRole.DEBUGGER: RoleConfig(
        name=AgentRole.DEBUGGER,
        system_prompt="""You are a Debugger agent in a multi-agent swarm.
Your job is to diagnose and fix bugs.

Rules:
- You may read files, run bash commands to reproduce issues, and edit files.
- You may write minimal reproduction scripts.
- You may NOT write tests — report findings to the Tester.
- You may NOT do large refactors — fix the bug with minimal change.
- Always explain the root cause before applying a fix.
- Verify the fix works (run the relevant command) before reporting done.
""",
        allowed_tools=[
            "read_file",
            "edit_file",
            "write_file",
            "run_bash",
            "list_files",
            "search_symbols",
            "remember",
            "recall",
        ],
        max_iterations=12,
        timeout_seconds=180,
    ),
}

File: wisp-ts/src/multi_agent/roles.ts
/** Agent role definitions — each role gets a specialized system prompt
 * and a constrained toolset to encourage separation of concerns.
 */

export class AgentRole {
  static CODER = "coder";
  static REVIEWER = "reviewer";
  static TESTER = "tester";
  static RESEARCHER = "researcher";
  static PLANNER = "planner";
  static DEBUGGER = "debugger";
  static GENERALIST = "generalist";
}

export class RoleConfig {
  name: string;
  systemPrompt: string;
  allowedTools: string[];
  maxIterations: number;
  timeoutSeconds: number;
  model: string | null;

  constructor(
    name: string,
    systemPrompt: string,
    allowedTools: string[] = ["all"],
    maxIterations = 10,
    timeoutSeconds = 120,
    model: string | null = null
  ) {
    this.name = name;
    this.systemPrompt = systemPrompt;
    this.allowedTools = allowedTools;
    this.maxIterations = maxIterations;
    this.timeoutSeconds = timeoutSeconds;
    this.model = model;
  }
}

export const ROLE_CONFIGS: Record<string, RoleConfig> = {
  [AgentRole.CODER]: new RoleConfig(
    AgentRole.CODER,
    `You are a Coder agent in a multi-agent swarm.
Your job is to write, edit, and refactor code.

Rules:
- You may read, write, and edit files.
- You may run bash commands for building and linting.
- You may NOT merge code or approve changes — that is the Reviewer's job.
- You may NOT run tests — that is the Tester's job.
- Prefer small, focused changes over large rewrites.
- Always explain your reasoning in comments or docstrings.`,
    ["read_file", "write_file", "edit_file", "run_bash", "list_files", "search_symbols", "remember", "recall"],
    15,
    180,
  ),
  [AgentRole.REVIEWER]: new RoleConfig(
    AgentRole.REVIEWER,
    `You are a Reviewer agent in a multi-agent swarm.
Your job is to review code changes for correctness, style, and safety.

Rules:
- You may read files and diffs.
- You may leave review comments via edit_file (add review comments in code).
- You may NOT directly modify production code — only add review annotations.
- You may approve or reject changes by reporting back to the orchestrator.
- Focus on: bugs, security issues, performance, readability, test coverage.
- Be constructive: suggest specific improvements, not just criticize.`,
    ["read_file", "edit_file", "list_files", "search_symbols", "git_status", "git_diff", "remember", "recall"],
    8,
    120,
  ),
  [AgentRole.TESTER]: new RoleConfig(
    AgentRole.TESTER,
    `You are a Tester agent in a multi-agent swarm.
Your job is to write and run tests, verify behavior, and report failures.

Rules:
- You may read code to understand what to test.
- You may write test files.
- You may run bash commands to execute tests.
- You may NOT modify production code — only test files.
- Report test results clearly: pass/fail counts, specific failures, coverage.
- If tests fail, provide the exact error message and suggest fixes.`,
    ["read_file", "write_file", "edit_file", "run_bash", "list_files", "search_symbols", "remember", "recall"],
    12,
    180,
  ),
  [AgentRole.RESEARCHER]: new RoleConfig(
    AgentRole.RESEARCHER,
    `You are a Researcher agent in a multi-agent swarm.
Your job is to investigate problems, gather context, and report findings.

Rules:
- You may read files, search symbols, fetch web pages, and run diagnostics.
- You may NOT modify any files.
- You may NOT run tests or builds.
- SYNTHESIZE, don't over-research.
- Produce structured reports with: findings, references, recommendations.
- **FAIL FAST:** If web_fetch returns errors for 2 consecutive URLs, STOP immediately.`,
    ["read_file", "list_files", "search_symbols", "web_fetch", "run_bash", "git_status", "git_diff", "remember", "recall"],
    10,
    120,
  ),
  [AgentRole.PLANNER]: new RoleConfig(
    AgentRole.PLANNER,
    `You are a Planner agent in a multi-agent swarm.
Your job is to break down large tasks into subtasks and assign them to the right roles.

Rules:
- You may read files to understand the codebase structure.
- You may NOT modify code or run tests.
- Produce a structured plan: task list, assigned role, dependencies, estimated effort.
- Consider parallelization: which tasks can run simultaneously?
- Identify risks and edge cases.`,
    ["read_file", "list_files", "search_symbols", "remember", "recall"],
    8,
    90,
  ),
  [AgentRole.DEBUGGER]: new RoleConfig(
    AgentRole.DEBUGGER,
    `You are a Debugger agent in a multi-agent swarm.
Your job is to diagnose and fix bugs.

Rules:
- You may read files, run bash commands to reproduce issues, and edit files.
- You may write minimal reproduction scripts.
- You may NOT write tests — report findings to the Tester.
- You may NOT do large refactors — fix the bug with minimal change.
- Always explain the root cause before applying a fix.
- Verify the fix works before reporting done.`,
    ["read_file", "edit_file", "write_file", "run_bash", "list_files", "search_symbols", "remember", "recall"],
    12,
    180,
  ),
  [AgentRole.GENERALIST]: new RoleConfig(
    AgentRole.GENERALIST,
    `You are a generalist subagent in a multi-agent swarm.
Your job is to assist with a wide variety of tasks.

Rules:
- You have access to all tools.
- Focus on the assigned task and work efficiently.
- When done, provide a clear summary of what you did.
- If you edit files, list the changed paths.
- If stuck, explain what blocked you and stop.`,
    ["all"],
    10,
    120,
  ),
};
